sentry-android-replay/src/main/java/io/sentry/android/replay/util/Nodes.kt

```diff
@@ -0,0 +1,127 @@
+ * That's not going to cover all cases, but probably 90%.
+ *
+ * We also add special proguard rules to keep the `Painter` class names and their `painter` member.
+ */
+internal fun LayoutNode.findPainter(): Painter? {
+    val modifierInfos = getModifierInfo()
+    for (index in modifierInfos.indices) {
+        val modifier = modifierInfos[index].modifier
+        if (modifier::class.java.name.contains("Painter")) {
+            return try {
+                modifier::class.java.getDeclaredField("painter")
+                    .apply { isAccessible = true }
+                    .get(modifier) as? Painter
+            } catch (e: Throwable) {
+                null
+            }
+        }
+    }
+    return null
+}
+
+/**
+ * We heuristically check the known classes that are coming from local assets usually:
+ * [androidx.compose.ui.graphics.vector.VectorPainter]
+ * [androidx.compose.ui.graphics.painter.ColorPainter]
+ * [androidx.compose.ui.graphics.painter.BrushPainter]
+ *
+ * In theory, [androidx.compose.ui.graphics.painter.BitmapPainter] can also come from local assets,
+ * but it can as well come from a network resource, so we preemptively redact it.
+ */
+internal fun Painter.isRedactable(): Boolean {
+    val className = this::class.java.name
+    return !className.contains("Vector") &&
+        !className.contains("Color") &&
+        !className.contains("Brush")
+}
+
+/**
+ * Converts from [androidx.compose.ui.geometry.Rect] to [android.graphics.Rect].
+ */
+internal fun androidx.compose.ui.geometry.Rect.toRect(): Rect {
+    return Rect(left.toInt(), top.toInt(), right.toInt(), bottom.toInt())
+}
+
+internal data class TextAttributes(val color: Color?, val hasFillModifier: Boolean)
+
+/**
+ * This method is necessary to redact text in Compose.
+ *
+ * We heuristically look up for classes that have a [Text] modifier, usually they all have a
+ * `Text` string in their name, e.g. TextStringSimpleElement or TextAnnotatedStringElement. We then
+ * get the color from the modifier, to be able to redact it with the correct color.
+ *
+ * We also look up for classes that have a [Fill] modifier, usually they all have a `Fill` string in
+ * their name, e.g. FillElement. This is necessary to workaround a Compose bug where single-line
+ * text composable without a `fill` modifier still thinks that there's one and wrongly calculates
+ * horizontal position.
+ *
+ * We also add special proguard rules to keep the `Text` class names and their `color` member.
+ */
+internal fun LayoutNode.findTextAttributes(): TextAttributes {
+    val modifierInfos = getModifierInfo()
+    var color: Color? = null
+    var hasFillModifier = false
+    for (index in modifierInfos.indices) {
+        val modifier = modifierInfos[index].modifier
+        val modifierClassName = modifier::class.java.name
+        if (modifierClassName.contains("Text")) {
+            color = try {
+                (
+                    modifier::class.java.getDeclaredField("color")
```

Comment:
maybe it makes sense to cache the field outside of `findTextAttributes()`

Reply:
yep, I thought about the same, but the problem here is that the holding class is actually private, so we'd have to access it via reflection too (`Class.forName`). And in this case we actually don't know what the class is until we start iterating over the modifiers list (there can be multiple classes holding this field), so we'd have to cache the field for each possible class, and the gains are questionable in this case.

in my tests it was quite fast (like 0.05ms on average for this method to execute), which resulted in the entire tree being parse within 4.5ms for an average screen. I think I'd rather focus on making it work via bytecode manipulation, so we wipe this entirely, wdyt?

```
E_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.13 ms
2024-10-08 17:53:47.663  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.664  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.09 ms
2024-10-08 17:53:47.664  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.03 ms
2024-10-08 17:53:47.665  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.666  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.12 ms
2024-10-08 17:53:47.667  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.06 ms
2024-10-08 17:53:47.667  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.03 ms
2024-10-08 17:53:47.671  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.19 ms
2024-10-08 17:53:47.672  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.673  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.22 ms
2024-10-08 17:53:47.674  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.09 ms
2024-10-08 17:53:47.674  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.675  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.07 ms
2024-10-08 17:53:47.675  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.676  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.678  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.06 ms
2024-10-08 17:53:47.679  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.13 ms
2024-10-08 17:53:47.680  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.680  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.681  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.681  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.683  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.07 ms
2024-10-08 17:53:47.685  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.16 ms
2024-10-08 17:53:47.685  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.686  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.686  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.687  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.689  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.06 ms
2024-10-08 17:53:47.690  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.10 ms
2024-10-08 17:53:47.690  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.691  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.691  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.691  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.693  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.08 ms
2024-10-08 17:53:47.694  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.03 ms
2024-10-08 17:53:47.694  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.695  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.10 ms
2024-10-08 17:53:47.696  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.696  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.03 ms
2024-10-08 17:53:47.696  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.697  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.699  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.08 ms
2024-10-08 17:53:47.700  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.10 ms
2024-10-08 17:53:47.700  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.701  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.701  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.702  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.704  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.07 ms
2024-10-08 17:53:47.704  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.10 ms
2024-10-08 17:53:47.705  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.705  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.706  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.706  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.708  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.06 ms
2024-10-08 17:53:47.709  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.12 ms
2024-10-08 17:53:47.710  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.03 ms
2024-10-08 17:53:47.711  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.09 ms
2024-10-08 17:53:47.711  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.711  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.714  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.08 ms
2024-10-08 17:53:47.715  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.11 ms
2024-10-08 17:53:47.715  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.716  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.716  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.716  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.719  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.08 ms
2024-10-08 17:53:47.720  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.11 ms
2024-10-08 17:53:47.720  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.721  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.721  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.721  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.724  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.07 ms
2024-10-08 17:53:47.724  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.11 ms
2024-10-08 17:53:47.725  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.725  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.726  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.726  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.728  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.07 ms
2024-10-08 17:53:47.729  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.13 ms
2024-10-08 17:53:47.729  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.730  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.730  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.730  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.01 ms
2024-10-08 17:53:47.732  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.07 ms
2024-10-08 17:53:47.733  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.11 ms
2024-10-08 17:53:47.733  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.734  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.04 ms
2024-10-08 17:53:47.734  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.02 ms
2024-10-08 17:53:47.734  7521-7521  TIME_IMAGE              au.com.shiftyjelly.pocketcasts       E  0.00 ms
2024-10-08 17:53:47.736  7521-7521  TIME_TEXT               au.com.shiftyjelly.pocketcasts       E  0.07 ms
```